File: polocloud-suite/src/main/java/dev/httpmarco/polocloud/suite/commands/impl/InfoCommand.java
package dev.httpmarco.polocloud.suite.commands.impl;

import com.sun.management.OperatingSystemMXBean;
import dev.httpmarco.polocloud.suite.PolocloudSuite;
import dev.httpmarco.polocloud.suite.commands.Command;
import lombok.extern.log4j.Log4j2;
import org.jetbrains.annotations.NotNull;

import java.lang.management.ManagementFactory;
import java.text.DecimalFormat;
import java.time.Duration;

@Log4j2
public final class InfoCommand extends Command {

    private static final DecimalFormat FORMAT = new DecimalFormat("#.##");

    public InfoCommand() {
        super("info", "Show all information about your running cluster.");

        var translation = PolocloudSuite.instance().translation();

        defaultExecution(it -> {
            // todo
            log.info(PolocloudSuite.instance().translation().get("suite.command.info.header", 1));

            double usedMemory = usedMemory();
            double maxMemory = maxMemory();
            log.info(translation.get("suite.command.info.memory.current", FORMAT.format(usedMemory), Math.round((100 / maxMemory) * usedMemory)));
            log.info(translation.get("suite.command.info.memory.max",  FORMAT.format(maxMemory)));
            log.info(translation.get("suite.command.info.cpu", cpuUsage()));
            log.info(translation.get("suite.command.info.uptime", formatDuration(System.currentTimeMillis() - Long.parseLong(System.getProperty("polocloud.boot.time")))));
            log.info(translation.get("suite.command.info.java", System.getProperty("java.version")));
            log.info(translation.get("suite.command.info.clusterType", PolocloudSuite.instance().cluster().name()));
            log.info(translation.get("suite.command.info.maxQueue", PolocloudSuite.instance().config().local().maxQueueProcesses()));
            log.info(translation.get("suite.command.info.idleShutdown", PolocloudSuite.instance().config().local().processTerminationIdleSeconds()));
        });
    }

    public double usedMemory() {
        var runtime = Runtime.getRuntime();
        var usedBytes = runtime.totalMemory() - runtime.freeMemory();

        return calculateMemory(usedBytes);
    }

    public double maxMemory() {
        var runtime = Runtime.getRuntime();
        var maxBytes = runtime.maxMemory();

        return calculateMemory(maxBytes);
    }

    public double cpuUsage() {
        OperatingSystemMXBean osBean = (OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean();
        double load = osBean.getCpuLoad();
        if (load < 0) return -1;
        return Math.round(load * 10000.0) / 100.0;
    }

    private double calculateMemory(long bytes) {
        return bytes / 1024.0 / 1024.0;
    }

    public static @NotNull String formatDuration(long millis) {
        Duration duration = Duration.ofMillis(millis);

        long days = duration.toDays();
        duration = duration.minusDays(days);

        long hours = duration.toHours();
        duration = duration.minusHours(hours);

        long minutes = duration.toMinutes();
        duration = duration.minusMinutes(minutes);

        long seconds = duration.getSeconds();
        StringBuilder sb = new StringBuilder();

        if (days > 0) sb.append(days).append("d ");
        if (hours > 0 || days > 0) sb.append(hours).append("h ");
        if (minutes > 0 || hours > 0 || days > 0) sb.append(minutes).append("m ");
        if (seconds > 0 || minutes > 0 || hours > 0 || days > 0) sb.append(seconds).append("s ");

        return sb.toString();
    }
}

File: agent/src/main/java/dev/httpmarco/polocloud/agent/runtime/local/terminal/commands/impl/InfoCommand.kt
package dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.impl

import com.sun.management.OperatingSystemMXBean
import dev.httpmarco.polocloud.agent.Agent
import dev.httpmarco.polocloud.agent.logger
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.Command
import java.lang.management.ManagementFactory
import java.time.Duration
import kotlin.math.roundToInt

class InfoCommand : Command("info", "Used to display information about the agent") {

    val osBean: OperatingSystemMXBean = ManagementFactory.getOperatingSystemMXBean() as OperatingSystemMXBean

    init {

        // call cpuUsage() to ensure the bean is initialized -> this is necessary because the bean might not be initialized yet when the command is executed
        cpuUsage()

        defaultExecution {
            logger.info("Current &3agent &7information&8:")
            logger.info("  &8- &7Uptime&8: &f${formatDuration(System.currentTimeMillis() - System.getProperty("polocloud.lifecycle.boot-time").toLong())}")
            logger.info("  &8- &7Cluster type&8: &f${Agent.instance.runtime.javaClass.simpleName}")
            logger.info("  &8- &7Java version&8: &f${System.getProperty("java.version")}")
            logger.info("  &8- &7Cpu usage&8: &f${cpuUsage()}%")
            logger.info("  &8- &7Used memory&8: &f${usedMemory()}MB")
            logger.info("  &8- &7Max memory&8: &f${maxMemory()}MB")
        }
    }

    private fun cpuUsage(): Double {
        val load = osBean.cpuLoad

        if (load < 0) {
            return -1.0
        }

        return (load * 10000.0).roundToInt() / 100.0
    }

    private fun usedMemory(): Double {
        val runtime = Runtime.getRuntime()
        val usedBytes = runtime.totalMemory() - runtime.freeMemory()

        return calculateMemory(usedBytes)
    }

    private fun maxMemory(): Double {
        val runtime = Runtime.getRuntime()
        val maxBytes = runtime.maxMemory()

        return calculateMemory(maxBytes)
    }

    private fun calculateMemory(bytes: Long): Double {
        return bytes / 1024.0 / 1024.0
    }

    fun formatDuration(millis: Long): String {
        var duration = Duration.ofMillis(millis)

        val days = duration.toDays()
        duration = duration.minusDays(days)

        val hours = duration.toHours()
        duration = duration.minusHours(hours)

        val minutes = duration.toMinutes()
        duration = duration.minusMinutes(minutes)

        val seconds = duration.getSeconds()
        val sb = StringBuilder()

        if (days > 0) sb.append(days).append("d ")
        if (hours > 0 || days > 0) sb.append(hours).append("h ")
        if (minutes > 0 || hours > 0 || days > 0) sb.append(minutes).append("m ")
        if (seconds > 0 || minutes > 0 || hours > 0 || days > 0) sb.append(seconds).append("s ")

        return sb.toString()
    }
}